### PR TITLE
Better debug logs for borrowck constraint graph

### PR DESCRIPTION
**Pull request by b-naber**
It's really cumbersome to work with `RegionVar`s when trying to debug borrowck code or when trying to understand how the borrowchecker works. This PR collects some region information (behind `cfg(debug_assertions)`) for created `RegionVar`s (NLL region vars, this PR doesn't touch canonicalization) and prints the nodes and edges of the strongly connected constraints graph using representatives that use that region information (either lifetime names, locations in MIR or spans).

### CODE REVIEW

**Comment by Nilstrieb** [comment]
This adds a huge amount of `#[cfg]`ed code, which isn't very good. I think most of this can be replaced with `cfg!(debug_assertions)` or other control-flow based conditional execution instead, which is way easier to maintain.
With something like
```rust
fn with_reg_var_to_origin(&self, f: impl FnOnce(RefCell<FxHashMap<ty::RegionVid, RegionCtxt>>))
```
that calls the closure when `debug_assertions` are enabled and is a noop when they aren't, the `#[cfg]` attributes could be contained to a minimum (one for the field, one for `new` and one for something like the function above).

**Comment by jackh726** [comment]
Can we try to add this without all the `cfg`s? And measure perf? Then add as few `cfg`s as possible to not regress perf?

**Comment by b-naber** [comment]
> Can we try to add this without all the `cfg`s? And measure perf? Then add as few `cfg`s as possible to not regress perf?

@jackh726 Sorry, didn't immediately have time when I saw your reply and then forgot that you posted.

**Comment by jackh726** [comment]
Agh sorry for the delay. Can you rebase and remove the cargo.lock changes?

**Comment by jackh726** [comment]
Agh

**Review by jackh726 — CHANGES_REQUESTED** [review]
Okay, so perf isn't that bad, but probably worth doing a few small things to reduce the impact in non-debug mode.

First thing I would do is split out the trivial changes into a separate PR. This would include debug deriving, adding instrument calls, and maybe the addition of BorrowckInferCtxt as a thin wrapper around InferCtxt.

After that, one easy thing to do would be to have next_nll_region_var take a function returning RegionCtxt, which only gets called in debug mode.

**Comment by jackh726** [comment]
@b-naber are you still interested in this PR? I could potentially take it over if you're unable to continue working on it.

**Comment by b-naber** [comment]
@jackh726 Sorry for completely neglecting this PR, I'll try to finish this.

Before splitting this into other PRs, can we run perf again (I've added back some cfg statements and changed `next_nll_region_var` to take a function as you suggested)?

**Comment by jackh726** [comment]
No worries.

Yes of course.

@bors try @rust-timer queue

**Comment by jackh726** [comment]
Okay good. Can we do one more with the cfgs removed?

**Comment by b-naber** [comment]
Forgot that I also should be able to request a perf run, let's see:

@bors try @rust-timer queue

**Comment by jackh726** [comment]
Okay, unless there's anything else you think you want to try, I think we land this as-is. How's that sound @b-naber?

**Comment by b-naber** [comment]
@jackh726 Don't have anything else to try.

File: compiler/rustc_borrowck/src/renumber.rs [hunk 1]
```diff
@@ -29,27 +31,68 @@ pub fn renumber_mir<'tcx>(
...
+    get_ctxt_fn: F,
+) -> T
 where
     T: TypeFoldable<'tcx>,
+    F: Fn() -> RegionCtxt,
 {
     infcx.tcx.fold_regions(value, |_region, _depth| {
         let origin = NllRegionVariableOrigin::Existential { from_forall: false };
-        infcx.next_nll_region_var(origin)
+        infcx.next_nll_region_var(origin, || get_ctxt_fn())
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/renumber.rs:46]: Can just pass `get_ctxt_fn` right?

**Review by jackh726 — APPROVED** [review]
Just a couple nits. r=me with or without them

File: compiler/rustc_borrowck/src/renumber.rs (hunk 1 above)
Comment (b-naber) [reply compiler/rustc_borrowck/src/renumber.rs:46]: No, `get_ctxt_fn` is captured in `fold_regions` which is `FnMut`, so we can't move there.

**Pull request merged**